libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,354 @@
+		{"cancelRequest", [](auto, auto) {/*nothing for now as we are synchronous */} },
+		{"exit", [this](auto, auto) { m_exitRequested = true; }},
+		{"initialize", bind(&LanguageServer::handleInitialize, this, _1, _2)},
+		{"initialized", [](auto, auto) {} },
+		{"shutdown", [this](auto, auto) { m_shutdownRequested = true; }},
+		{"textDocument/didChange", bind(&LanguageServer::handleTextDocumentDidChange, this, _1, _2)},
+		{"textDocument/didClose", [](auto, auto) {/*nothing for now*/}},
+		{"textDocument/didOpen", bind(&LanguageServer::handleTextDocumentDidOpen, this, _1, _2)},
+		{"workspace/didChangeConfiguration", bind(&LanguageServer::handleWorkspaceDidChangeConfiguration, this, _1, _2)},
+	},
+	m_fileReader{"/" /* base path */},
+	m_compilerStack{bind(&FileReader::readFile, ref(m_fileReader), _1, _2)}
+{
+}
+
+optional<SourceLocation> LanguageServer::parsePosition(
+	string const& _sourceUnitName,
+	Json::Value const& _position
+) const
+{
+	if (!m_fileReader.sourceCodes().count(_sourceUnitName))
+		return nullopt;
+
+	if (optional<LineColumn> lineColumn = parseLineColumn(_position))
+		if (optional<int> const offset = CharStream::translateLineColumnToPosition(
+			m_fileReader.sourceCodes().at(_sourceUnitName),
+			*lineColumn
+		))
+			return SourceLocation{*offset, *offset, make_shared<string>(_sourceUnitName)};
+	return nullopt;
+}
+
+optional<SourceLocation> LanguageServer::parseRange(string const& _sourceUnitName, Json::Value const& _range) const
+{
+	if (!_range.isObject())
+		return nullopt;
+	optional<SourceLocation> start = parsePosition(_sourceUnitName, _range["start"]);
+	optional<SourceLocation> end = parsePosition(_sourceUnitName, _range["end"]);
+	if (!start || !end)
+		return nullopt;
+	solAssert(*start->sourceName == *end->sourceName);
+	start->end = end->end;
+	return start;
+}
+
+Json::Value LanguageServer::toRange(SourceLocation const& _location) const
+{
+	if (!_location.hasText())
+		return toJsonRange({}, {});
+
+	solAssert(_location.sourceName, "");
+	CharStream const& stream = m_compilerStack.charStream(*_location.sourceName);
+	LineColumn start = stream.translatePositionToLineColumn(_location.start);
+	LineColumn end = stream.translatePositionToLineColumn(_location.end);
+	return toJsonRange(start, end);
+}
+
+Json::Value LanguageServer::toJson(SourceLocation const& _location) const
+{
+	solAssert(_location.sourceName);
+	Json::Value item = Json::objectValue;
+	item["uri"] = sourceUnitNameToClientPath(*_location.sourceName);
+	item["range"] = toRange(_location);
+	return item;
+}
+
+string LanguageServer::clientPathToSourceUnitName(string const& _path) const
+{
+	string path = _path;
+	if (path.find("file://") == 0)
+		path = path.substr(7);
+
+	return m_fileReader.cliPathToSourceUnitName(path);
+}
+
+string LanguageServer::sourceUnitNameToClientPath(string const& _sourceUnitName) const
+{
+	return "file://" + (m_fileReader.basePath() / _sourceUnitName).generic_string();
```

Comment:
What is important that this is the opposite of `clientPathToSourceUnitName`. The stuff inside FileReader is extremely complicated. Note that we can assume that the URIs we get from the client are fully normalized.

Reply:
True, the normalization itself is complicated but all the other stuff I mentioned above is not and do (or will) need it in the LSP anyway to get consistent behavior. I'd just use `FileReader` despite the normalization rather than duplicate it here.